File: c++/003/003.cpp
#include <iostream>
#include <vector>

std::vector<unsigned long long> primeFactors(unsigned long long n);

int main() {
  std::cout << primeFactors(600851475143).back() << std::endl;

  return 0;
}

std::vector<unsigned long long> primeFactors(unsigned long long n) {
  bool found = false;

  unsigned long long v = 2;

  for (unsigned long long i = 2; i < n - 1; i++) {
    if (n % i == 0) {
      found = true;

      v = i;

      break;
    }
  }

  std::vector<unsigned long long> result;

  if (found) {
    result.push_back(v);

    std::vector<unsigned long long> moreFactors = primeFactors(n / v);

    result.insert(result.end(), moreFactors.begin(), moreFactors.end());
  } else {
    result.push_back(n);
  }

  return result;
}

File: go/003/003.go
package main

import "fmt"

func main() {
	factors := primeFactors(600851475143)

	fmt.Println(factors[len(factors)-1])
}

func primeFactors(x uint64) []uint64 {
	var found bool = false

	var v uint64 = 2

	var i uint64

	for i = 2; i < x-1; i++ {
		if x%i == 0 {
			found = true

			v = i

			break
		}
	}

	var result []uint64

	if found {
		result = append(result, v)

		moreFactors := primeFactors(x / v)

		result = concat(result, moreFactors)
	} else {
		result = append(result, x)
	}

	return result
}

func concat(a, b []uint64) []uint64 {
	c := make([]uint64, len(a)+len(b))

	copy(c[:len(a)], a)

	copy(c[len(a):], b)

	return c
}
